### PR TITLE
Feature/dynamic types

### PR DESCRIPTION
**Pull request by LuisGP**
- [x] Added support for dynamic types 1.1 (minimals checking)
- [x] Added TypeObject and TypeIdentifier generation
- [x] Conversion functions between dynamic types and TypeObject/TypeIdentifier

### CODE REVIEW

File: examples/C++/DynamicHelloWorldExample/HelloWorldPublisher.cpp [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+    , m_DynType(nullptr)
+    , m_dynamic(false)
+{
+    m_Hello = nullptr;
+}
+
+bool HelloWorldPublisher::init(bool dynamic)
+{
+    m_dynamic = dynamic;
+    if (dynamic)
```
Comment (richiware) [inline examples/C++/DynamicHelloWorldExample/HelloWorldPublisher.cpp]: This example is to show how to create a dynamic data. Erase the `else` part.

File: examples/C++/DynamicHelloWorldExample/HelloWorldPublisher.cpp [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+    }
+    else
+    {
+        m_Hello = new HelloWorld;
+        m_Hello->index(0);
+        m_Hello->message("HelloWorld");
+    }
+
+    ParticipantAttributes PParam;
+    PParam.rtps.builtin.use_SIMPLE_RTPSParticipantDiscoveryProtocol = true;
```
Comment (richiware) [inline examples/C++/DynamicHelloWorldExample/HelloWorldPublisher.cpp]: Simplify configuration. Only change domainId and name.

File: examples/C++/DynamicHelloWorldExample/HelloWorldPublisher.cpp [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    else
+    {
+        Domain::registerType(mp_participant, &m_type);
+    }
+
+    //CREATE THE PUBLISHER
+    PublisherAttributes Wparam;
+    Wparam.topic.topicKind = NO_KEY;
+    Wparam.topic.topicDataType = "HelloWorld";
+    Wparam.topic.topicName = "DynamicHelloWorldTopic";
```
Comment (richiware) [inline examples/C++/DynamicHelloWorldExample/HelloWorldPublisher.cpp]: Use same topicName as normal HelloWorldExample. Doing this both examples can interoperate.

File: examples/C++/DynamicHelloWorldExample/HelloWorldPublisher.cpp [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    Wparam.topic.topicDiscoveryKind = MINIMAL;
+    Wparam.topic.historyQos.kind = KEEP_LAST_HISTORY_QOS;
```
Comment (richiware) [inline examples/C++/DynamicHelloWorldExample/HelloWorldPublisher.cpp]: Minimize configurate. Remove to use default values.

File: examples/C++/DynamicHelloWorldExample/HelloWorldPublisher.cpp [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+    }
+    else if (info.rtps.m_status == DROPPED_RTPSPARTICIPANT)
+    {
+        std::cout << "Participant " << info.rtps.m_RTPSParticipantName << " dropped" << std::endl;
+    }
+}
+
+void HelloWorldPublisher::runThread(uint32_t samples, uint32_t sleep)
+{
+    if (samples == 0)
```
Comment (richiware) [inline examples/C++/DynamicHelloWorldExample/HelloWorldPublisher.cpp]: Unify in only one while statement.

File: examples/C++/DynamicHelloWorldExample/HelloWorldSubscriber.cpp [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+    , mp_subscriber(nullptr)
+    , m_DynType(nullptr)
+{
+}
+
+bool HelloWorldSubscriber::init(bool dynamic)
+{
+    ParticipantAttributes PParam;
+    PParam.rtps.builtin.use_SIMPLE_RTPSParticipantDiscoveryProtocol = true;
+    PParam.rtps.builtin.use_SIMPLE_EndpointDiscoveryProtocol = true;
```
Comment (richiware) [inline examples/C++/DynamicHelloWorldExample/HelloWorldSubscriber.cpp]: Minimize configuration. Use default values, except in domainId and name.

File: examples/C++/DynamicHelloWorldExample/HelloWorldSubscriber.cpp [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    PParam.rtps.builtin.leaseDuration = c_TimeInfinite;
+    PParam.rtps.setName("DynHelloWorld_sub");
+    mp_participant = Domain::createParticipant(PParam, (ParticipantListener*)&m_part_list);
+    if(mp_participant==nullptr)
+        return false;
+
+    //REGISTER THE TYPE
+    m_dynamic = dynamic;
+    m_listener.m_dynamic = dynamic;
+    if (dynamic)
```
Comment (richiware) [inline examples/C++/DynamicHelloWorldExample/HelloWorldSubscriber.cpp]: This example is to show the usage of dynamic types. Remove the `else` part.

File: examples/C++/DynamicHelloWorldExample/HelloWorldSubscriber.cpp [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+    {
+        // Create basic types
+        DynamicTypeBuilder_ptr created_type_ulong = DynamicTypeBuilderFactory::GetInstance()->CreateUint32Builder();
```
Comment (richiware) [inline examples/C++/DynamicHelloWorldExample/HelloWorldSubscriber.cpp]: Join these two lines in the section of add members.

File: examples/C++/DynamicHelloWorldExample/HelloWorldSubscriber.cpp [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    //CREATE THE SUBSCRIBER
+    SubscriberAttributes Rparam;
+    Rparam.topic.topicKind = NO_KEY;
+    Rparam.topic.topicDataType = "HelloWorld";
+    Rparam.topic.topicName = "DynamicHelloWorldTopic";
+    Rparam.topic.topicDiscoveryKind = MINIMAL;
+    //Rparam.topic.topicDiscoveryKind = NO_CHECK;
+    Rparam.topic.historyQos.kind = KEEP_LAST_HISTORY_QOS;
+    Rparam.topic.historyQos.depth = 30;
+    Rparam.topic.resourceLimitsQos.max_samples = 50;
```
Comment (richiware) [inline examples/C++/DynamicHelloWorldExample/HelloWorldSubscriber.cpp]: Minimize configuration trying to use default values.

File: examples/C++/DynamicHelloWorldExample/HelloWorldSubscriber.cpp [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+    {
+        m_listener.m_Hello = new HelloWorld;
+        Domain::registerType(mp_participant,&m_type);
+    }
+
+    //CREATE THE SUBSCRIBER
+    SubscriberAttributes Rparam;
+    Rparam.topic.topicKind = NO_KEY;
+    Rparam.topic.topicDataType = "HelloWorld";
+    Rparam.topic.topicName = "DynamicHelloWorldTopic";
```
Comment (richiware) [inline examples/C++/DynamicHelloWorldExample/HelloWorldSubscriber.cpp]: Change topic name. Set the names used in normal helloworld example.

File: fastrtpsgen/src/com/eprosima/fastrtps/idl/templates/CMakeLists.stg [hunk 11]
```diff
@@ -34,10 +34,7 @@ $solution.projects : {$pub_sub_execs(project=it, libraries=solution.libraries)$}
 pub_sub_execs(project, libraries) ::= <<
 
 message(STATUS "Configuring $project.name$...")
-add_library($project.name$_lib $project.commonSrcFiles : {$it$}; separator=" "$)
-target_link_libraries($project.name$_lib $solution.libraries : {$it$}; separator=" "$)
+add_executable($project.name$ $project.commonSrcFiles : {$it$}; separator=" "$ $project.projectSrcFiles : {$it$}; separator=" "$)
```
Comment (richiware) [inline fastrtpsgen/src/com/eprosima/fastrtps/idl/templates/CMakeLists.stg]: I believe a rebase remove the correct code. Use the develop branch version.

File: include/fastrtps/Domain.h [hunk 12]
```diff
@@ -20,7 +20,8 @@
 #ifndef DOMAIN_H_
 #define DOMAIN_H_
 
-#include "attributes/ParticipantAttributes.h"
+#include <fastrtps/attributes/ParticipantAttributes.h>
+#include <fastrtps/types/DynamicTypePtr.h>
```
Comment (richiware) [inline include/fastrtps/Domain.h]: This include is not necessary here.

File: include/fastrtps/Domain.h [hunk 13]
```diff
@@ -20,7 +20,7 @@
 #ifndef DOMAIN_H_
 #define DOMAIN_H_
 
-#include "attributes/ParticipantAttributes.h"
+#include <fastrtps/attributes/ParticipantAttributes.h>
```
Comment (richiware) [inline include/fastrtps/Domain.h]: Using quotes is faster for preprocessor. Mantain quotes
```suggestion
#include "attributes/ParticipantAttributes.h"
```

Comment (LuisGP) [reply include/fastrtps/Domain.h]: But that can cause problems when mocking, in this case, if ParticipantAttributes is mocked (I don't remember if it is the case)

Comment (richiware) [reply include/fastrtps/Domain.h]: Remember we found a solution when a class was mocking.

**Comment by LuisGP** [comment]
DynamicHelloWorld was publishing at domain 5 instead 0 and probably, the minimal type checking wasn't helping neither.

**Pull request merged**